### PR TITLE
Update commands with resource builders to use external versions

### PR DESCRIPTION
**Pull request by juanvallejo**
Updates any remaining commands that use resource builders to deal with external versions

### CODE REVIEW

File: pkg/oc/cli/debug/debug.go [hunk 1]
```diff
@@ -256,6 +254,7 @@ func (o *DebugOptions) Complete(cmd *cobra.Command, f kcmdutil.Factory, args []s
 
 	o.Builder = f.NewBuilder
 
+	// TODO: not sure what to do with this helper - new duplicate for external versions?
 	o.AddEnv, o.RemoveEnv, err = utilenv.ParseEnv(envArgs, nil)
```
Comment (juanvallejo) [inline pkg/oc/cli/debug/debug.go]: @deads2k @soltysh still not sure what to do with this helper, duplicate it to deal with externals? - it depends on other env utils

File: pkg/oc/cli/debug/debug.go [hunk 2]
```diff
@@ -562,6 +554,7 @@ func (o *DebugOptions) getContainerImageCommand(pod *kapi.Pod, container *kapi.C
 		image, _ = o.getContainerImageViaImageStreamImport(container)
 	}
 
+	// TODO: DockerImageMetadata is a RawExtension in the external api - how to access config?
 	if image == nil || image.DockerImageMetadata.Config == nil {
```
Comment (juanvallejo) [inline pkg/oc/cli/debug/debug.go]: @deads2k @soltysh Not sure how to access config here - DockerImageMetadata is a RawExtension in the external api

File: pkg/oc/cli/debug/debug.go (hunk 1 above)
Comment (soltysh) [reply pkg/oc/cli/debug/debug.go]: See https://github.com/openshift/origin/pull/20382/files#r204804974

File: pkg/oc/cli/debug/debug.go [hunk 3]
```diff
@@ -562,16 +569,24 @@ func (o *DebugOptions) getContainerImageCommand(pod *kapi.Pod, container *kapi.C
 		image, _ = o.getContainerImageViaImageStreamImport(container)
 	}
 
-	if image == nil || image.DockerImageMetadata.Config == nil {
+	// TODO: find a way to do this with the external api
+	// convert image to internal in order to extract its imagemetadata config
+	internalImage := &internalimage.Image{}
+	if err := oapiv1.Convert_v1_Image_To_image_Image(image, internalImage, nil); err != nil {
```
Comment (juanvallejo) [inline pkg/oc/cli/debug/debug.go]: @deads2k Is there a more elegant way to access an image's DockerImageMetadata.Config field? The external api stores the `DockerImageMetadata` field as a RawExtension

File: pkg/oc/cli/debug/debug.go [hunk 4]
```diff
@@ -39,62 +39,21 @@ import (
 	"k8s.io/kubernetes/pkg/kubectl/util/term"
 	"k8s.io/kubernetes/pkg/util/interrupt"
 
+	appsapiv1 "github.com/openshift/api/apps/v1"
 	appsv1 "github.com/openshift/api/apps/v1"
```
Comment (soltysh) [inline pkg/oc/cli/debug/debug.go]: You have duplicates b/c of different aliases.

File: pkg/oc/cli/debug/debug.go [hunk 5]
```diff
@@ -39,62 +39,21 @@ import (
...
+	imageapiv1 "github.com/openshift/api/image/v1"
 	imagev1 "github.com/openshift/api/image/v1"
```
Comment (soltysh) [inline pkg/oc/cli/debug/debug.go]: ditto

File: pkg/oc/cli/debug/debug.go [hunk 6]
```diff
@@ -39,62 +39,21 @@ import (
...
+	appsclientv1 "github.com/openshift/client-go/apps/clientset/versioned/typed/apps/v1"
+	imageclientv1 "github.com/openshift/client-go/image/clientset/versioned/typed/image/v1"
```
Comment (soltysh) [inline pkg/oc/cli/debug/debug.go]: `appsv1client` `imagev1client`

File: pkg/oc/cli/debug/debug.go [hunk 7]
```diff
@@ -39,62 +39,21 @@ import (
...
 	appsapi "github.com/openshift/origin/pkg/apps/apis/apps"
-	appsclientinternal "github.com/openshift/origin/pkg/apps/generated/internalclientset"
-	appsclient "github.com/openshift/origin/pkg/apps/generated/internalclientset/typed/apps/internalversion"
 	appsutil "github.com/openshift/origin/pkg/apps/util"
 	imageapi "github.com/openshift/origin/pkg/image/apis/image"
-	imageclientinternal "github.com/openshift/origin/pkg/image/generated/internalclientset"
-	imageclient "github.com/openshift/origin/pkg/image/generated/internalclientset/typed/image/internalversion"
+	internalimage "github.com/openshift/origin/pkg/image/apis/image"
+	oapiv1 "github.com/openshift/origin/pkg/image/apis/image/v1"
```
Comment (soltysh) [inline pkg/oc/cli/debug/debug.go]: `imageapiv1`

File: pkg/oc/cli/debug/debug.go [hunk 8]
```diff
@@ -256,11 +257,24 @@ func (o *DebugOptions) Complete(cmd *cobra.Command, f kcmdutil.Factory, args []s
 
 	o.Builder = f.NewBuilder
 
-	o.AddEnv, o.RemoveEnv, err = utilenv.ParseEnv(envArgs, nil)
+	// TODO: not sure what to do with this helper - new duplicate for external versions?
```
Comment (soltysh) [inline pkg/oc/cli/debug/debug.go]: Remove TODO, you're doing conversions in place below.

File: pkg/oc/cli/debug/debug.go (hunk 3 above)
Comment (soltysh) [reply pkg/oc/cli/debug/debug.go]: We'll have to live with it for now.

File: pkg/oc/cli/extract/extract.go [hunk 9]
```diff
@@ -131,9 +130,21 @@ func name(info *resource.Info) string {
 }
 
 func (o *ExtractOptions) Run() error {
+	r := o.Builder().
+		WithScheme(ocscheme.ReadingInternalScheme, ocscheme.ReadingInternalScheme.PrioritizedVersionsAllGroups()...).
```
Comment (soltysh) [inline pkg/oc/cli/extract/extract.go]: `scheme.Scheme`

File: pkg/oc/cli/rsh/rsh.go [hunk 10]
```diff
@@ -6,9 +6,11 @@ import (
 	"time"
 
 	"github.com/spf13/cobra"
+
 	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
```
Comment (soltysh) [inline pkg/oc/cli/rsh/rsh.go]: `corev1client`

File: pkg/oc/cli/rsync/pathspec.go [hunk 11]
```diff
@@ -96,7 +96,7 @@ func resolveResourceKindPath(f kcmdutil.Factory, path, namespace string) (string
 	}
 
 	r := f.NewBuilder().
-		WithScheme(ocscheme.ReadingInternalScheme).
+		WithScheme(ocscheme.ReadingInternalScheme, ocscheme.ReadingInternalScheme.PrioritizedVersionsAllGroups()...).
```
Comment (soltysh) [inline pkg/oc/cli/rsync/pathspec.go]: `scheme.Scheme`

File: pkg/oc/cli/secrets/basicauth.go [hunk 12]
```diff
@@ -3,19 +3,20 @@ package secrets
...
 	"errors"
 	"fmt"
-	"io"
 	"io/ioutil"
 
 	"github.com/spf13/cobra"
 
-	api "k8s.io/kubernetes/pkg/apis/core"
-	kcoreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+	coreapiv1 "k8s.io/api/core/v1"
```
Comment (soltysh) [inline pkg/oc/cli/secrets/basicauth.go]: `corev1`

File: pkg/oc/cli/secrets/basicauth.go [hunk 13]
```diff
@@ -3,19 +3,20 @@ package secrets
...
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
```
Comment (soltysh) [inline pkg/oc/cli/secrets/basicauth.go]: `corev1client`

File: pkg/oc/cli/secrets/basicauth.go [hunk 14]
```diff
@@ -81,15 +89,7 @@ func NewCmdCreateBasicAuthSecret(name, fullName string, f kcmdutil.Factory, stre
...
-			if len(kcmdutil.GetFlagString(c, "output")) != 0 {
-				secret, err := o.NewBasicAuthSecret()
-				kcmdutil.CheckErr(err)
-
-				kcmdutil.CheckErr(kcmdutil.PrintObject(c, secret, streams.Out))
-				return
-			}
-
-			if err := o.CreateBasicAuthSecret(); err != nil {
+			if err := o.Run(); err != nil {
```
Comment (soltysh) [inline pkg/oc/cli/secrets/basicauth.go]: Convert all these invocations to `kcmdutil.CheckErr(o.Complete|Validate|Run)`

File: pkg/oc/cli/secrets/dockercfg.go [hunk 15]
```diff
@@ -4,18 +4,18 @@ import (
...
-	"io"
-	"io/ioutil"
 	"strings"
 
-	api "k8s.io/kubernetes/pkg/apis/core"
-	kcoreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/origin/pkg/oc/util/ocscheme"
+	coreapiv1 "k8s.io/api/core/v1"
```
Comment (soltysh) [inline pkg/oc/cli/secrets/dockercfg.go]: `corev1`

File: pkg/oc/cli/secrets/dockercfg.go [hunk 16]
```diff
@@ -4,18 +4,18 @@ import (
...
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
```
Comment (soltysh) [inline pkg/oc/cli/secrets/dockercfg.go]: `corev1client`

File: pkg/oc/cli/secrets/dockercfg.go [hunk 17]
```diff
@@ -81,15 +92,7 @@ func NewCmdCreateDockerConfigSecret(name, fullName string, f kcmdutil.Factory, s
...
-			if len(kcmdutil.GetFlagString(c, "output")) != 0 {
-				secret, err := o.NewDockerSecret()
-				kcmdutil.CheckErr(err)
-
-				kcmdutil.CheckErr(kcmdutil.PrintObject(c, secret, streams.Out))
-				return
-			}
-
-			if err := o.CreateDockerSecret(); err != nil {
+			if err := o.Run(); err != nil {
```
Comment (soltysh) [inline pkg/oc/cli/secrets/dockercfg.go]: `kcmdutil.CheckErr(o....)` each time

File: pkg/oc/cli/secrets/new.go [hunk 18]
```diff
@@ -3,21 +3,22 @@ package secrets
...
 	"fmt"
-	"io"
 	"io/ioutil"
 	"os"
 	"path"
 	"strings"
 
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/origin/pkg/oc/util/ocscheme"
```
Comment (soltysh) [inline pkg/oc/cli/secrets/new.go]: Move down to origin's imports.

File: pkg/oc/cli/secrets/new.go [hunk 19]
```diff
@@ -3,21 +3,22 @@ package secrets
...
+	coreapiv1 "k8s.io/api/core/v1"
```
Comment (soltysh) [inline pkg/oc/cli/secrets/new.go]: `corev1`

File: pkg/oc/cli/secrets/new.go [hunk 20]
```diff
@@ -3,21 +3,22 @@ package secrets
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	kvalidation "k8s.io/apimachinery/pkg/util/validation"
-	kapi "k8s.io/kubernetes/pkg/apis/core"
-	kcoreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
```
Comment (soltysh) [inline pkg/oc/cli/secrets/new.go]: `corev1client`

File: pkg/oc/cli/secrets/new.go [hunk 21]
```diff
@@ -83,42 +91,28 @@ func NewCmdCreateSecret(name, fullName string, f kcmdutil.Factory, streams gener
 		Deprecated: "use oc create secret",
 		Hidden:     true,
 		Run: func(c *cobra.Command, args []string) {
-			if err := options.Complete(args, f); err != nil {
+			if err := o.Complete(args, f); err != nil {
 				kcmdutil.CheckErr(kcmdutil.UsageErrorf(c, err.Error()))
 			}
 
-			if err := options.Validate(); err != nil {
+			if err := o.Validate(); err != nil {
```
Comment (soltysh) [inline pkg/oc/cli/secrets/new.go]: `kcmdutil.CheckErr(o...)` each time

File: pkg/oc/cli/secrets/options.go [hunk 22]
```diff
@@ -3,16 +3,14 @@ package secrets
...
 	"os"
 	"strings"
 
+	corev1 "k8s.io/api/core/v1"
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/sets"
-	kapi "k8s.io/kubernetes/pkg/apis/core"
-	kcoreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+	kcoreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
```
Comment (soltysh) [inline pkg/oc/cli/secrets/options.go]: `corev1client`

File: pkg/oc/cli/secrets/options.go [hunk 23]
```diff
@@ -49,7 +45,7 @@ func (o *SecretOptions) Complete(f kcmdutil.Factory, args []string) error {
 	if err != nil {
 		return err
 	}
-	kubeClient, err := kcoreclient.NewForConfig(clientConfig)
+	kubeClient, err := kcoreclientv1.NewForConfig(clientConfig)
```
Comment (soltysh) [inline pkg/oc/cli/secrets/options.go]: Assign directly to `o.KubeCoreClient`. Btw, you can just name that var `KubeClient`, that's shorter :wink:

File: pkg/oc/cli/secrets/options.go [hunk 24]
```diff
@@ -88,9 +84,9 @@ func (o SecretOptions) Validate() error {
 }
 
 // GetServiceAccount Retrieve the service account object specified by the command
-func (o SecretOptions) GetServiceAccount() (*kapi.ServiceAccount, error) {
+func (o SecretOptions) GetServiceAccount() (*corev1.ServiceAccount, error) {
 	r := o.BuilderFunc().
-		WithScheme(ocscheme.ReadingInternalScheme).
+		WithScheme(ocscheme.ReadingInternalScheme, ocscheme.ReadingInternalScheme.PrioritizedVersionsAllGroups()...).
```
Comment (soltysh) [inline pkg/oc/cli/secrets/options.go]: `scheme.Scheme`

File: pkg/oc/cli/secrets/sshauth.go [hunk 25]
```diff
@@ -3,16 +3,17 @@ package secrets
...
 	"errors"
 	"fmt"
-	"io"
 	"io/ioutil"
 
-	api "k8s.io/kubernetes/pkg/apis/core"
-	kcoreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/origin/pkg/oc/util/ocscheme"
```
Comment (soltysh) [inline pkg/oc/cli/secrets/sshauth.go]: Move to origin's imports.

File: pkg/oc/cli/secrets/sshauth.go [hunk 26]
```diff
@@ -71,19 +81,11 @@ func NewCmdCreateSSHAuthSecret(name, fullName string, f kcmdutil.Factory, stream
 				kcmdutil.CheckErr(kcmdutil.UsageErrorf(c, err.Error()))
 			}
 
-			if err := o.Validate(); err != nil {
+			if err := o.Validate(args); err != nil {
```
Comment (soltysh) [inline pkg/oc/cli/secrets/sshauth.go]: `kcmdutil.CheckErr(o...` on each call

File: pkg/oc/cli/secrets/unlink_secret_from_object.go [hunk 27]
```diff
@@ -4,12 +4,14 @@ import (
 	"errors"
 	"fmt"
 
-	kapi "k8s.io/kubernetes/pkg/apis/core"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/origin/pkg/oc/util/ocscheme"
```
Comment (soltysh) [inline pkg/oc/cli/secrets/unlink_secret_from_object.go]: Move down to origin's imports.

File: pkg/oc/cli/secrets/dockercfg.go [hunk 28]
```diff
@@ -100,27 +95,31 @@ func NewCmdCreateDockerConfigSecret(name, fullName string, f kcmdutil.Factory, s
...
+func (o CreateDockerConfigOptions) Run() error {
 	secret, err := o.NewDockerSecret()
 	if err != nil {
 		return err
 	}
 
-	if _, err := o.SecretsInterface.Create(secret); err != nil {
-		return err
-	}
+	// TODO: sweep codebase removing implied --dry-run behavior when -o is specified
```
Comment (soltysh) [inline pkg/oc/cli/secrets/dockercfg.go]: @juanvallejo create an issue about it.

**Comment by soltysh** [comment]
Closed in favor of https://github.com/openshift/origin/pull/20514.

**Pull request closed without merging**